lib/app.js

```diff
@@ -1,476 +1,298 @@
-
-      if      (shell.test('-e', newTrufflePath)) truffleConfig = reqCwd.silent(newTrufflePath);
-      else if (shell.test('-e', oldTrufflePath)) truffleConfig = reqCwd.silent(oldTrufflePath);
-
-      this.network = '--network coverage';
-
-      // Coverage network opts specified: use port if declared
-      if (truffleConfig.networks && truffleConfig.networks.coverage) {
-        this.port = truffleConfig.networks.coverage.port || this.port;
-      } else {
-        // Put the coverage network in the existing config
-        if (!truffleConfig.networks) truffleConfig.networks = {};
-        truffleConfig.networks.coverage = coverageNetwork;
-        const configString = `module.exports = ${JSON.stringify(truffleConfig)}`;
-        fs.writeFileSync(`${this.coverageDir}/truffle-config.js`, configString);
-      }
+    this.skippedFolders = [];
+    this.skipFiles = config.skipFiles || [];
 
-      // Compile the contracts before instrumentation and preserve their ABI's.
-      // We will be stripping out access modifiers like view before we recompile
-      // post-instrumentation.
-      if (shell.test('-e', `${this.coverageDir}${this.buildDirPath}`)){
-        shell.rm('-Rf', `${this.coverageDir}${this.buildDirPath}`)
-      }
+    this.log = config.logger ? config.logger.log : console.log;
+    this.setLoggingLevel(config.silent);
 
-      this.runCompileCommand();
-      this.originalArtifacts = this.loadArtifacts();
-      shell.rm('-Rf', `${this.coverageDir}${this.buildDirPath}`);
+    this.gasLimit = 0xfffffffffff;
+    this.gasLimitString = "0xfffffffffff";
+    this.gasPrice = 0x01;
 
-    } catch (err) {
-      const msg = ('There was a problem generating the coverage environment: ');
-      this.cleanUp(msg + err);
-    }
+    this.istanbulReporter = config.istanbulReporter || ['html', 'lcov', 'text'];
+    this.ui = new UI(this.log);
   }
 
+  // --------------------------------------  Methods -----------------------------------------------
   /**
-   * For each contract except migrations.sol (or those in skipFiles):
-   * + Generate file path reference for coverage report
-   * + Load contract as string
-   * + Instrument contract
-   * + Save instrumented contract in the coverage environment folder where covered tests will run
-   * + Add instrumentation info to the coverage map
+   * Setup temp folder, write instrumented contracts to it and register them as coverage targets
    */
-  instrumentTarget() {
-    this.skipFiles = this.skipFiles.map(contract => `${this.coverageDir}/contracts/${contract}`);
-    this.skipFiles.push(`${this.coverageDir}/contracts/Migrations.sol`);
-
-    const instrumentedFiles = [];
+  instrument() {
     let currentFile;
+    let started = false;
+    let skipped = [];
+
     try {
-      shell.ls(`${this.coverageDir}/contracts/**/*.sol`).forEach(file => {
+      this.sanityCheckContext();
+      this.registerSkippedItems();
+      this.generateEnvelope();
```

Comment:
This needs to be abstracted out of `instrument()` because in buidler we can instrument in-flight, e.g. without writing to a separate folder.

Reply:
#406